### DISCUSSION TITLE
how to connect two devices with appium server

### DISCUSSION THREAD

**Discussion by 9iuzhongxian** [appium/appium#18890] [Q&A]
hey guys,recently I encountered some difficulties.I want to start two appium server with different port and the scrennshot like this 
![image](https://github.com/appium/appium/assets/44670292/66b93ac7-0c14-4b91-9450-c97729965f71)
![image](https://github.com/appium/appium/assets/44670292/a198d6ac-ecd5-458e-8929-cc930a2fecb6)
when i want to connect these two server,and i want to asscociate these servers with  two emulator,and so I can via the client connect the emulator,but when i serach how to connect the two devices,the search engine give me answers like this 
`appium -p <server-listening-port> -U<device-name> -bp<device-port>`
i used the appium2,and when i try this 
![image](https://github.com/appium/appium/assets/44670292/53697a14-dba3-4a27-9060-890ad15664b3)
![image](https://github.com/appium/appium/assets/44670292/8881e0a8-7c88-4301-8670-0a5116d34c17)
so when i tried and failed ,i use the desiredcap with
`deviceName="127.0.0.1:62001"`
when i use python to connect and manipulate
`
import time

from appium import  webdriver
from appium.webdriver.common.appiumby import AppiumBy
cap = {
    "platformName": "Android",
    "automationName": "uiautomator2",
    "appPackage": "com.android.launcher3",
    "appActivaty": ".launcher3.Launcher",
    "language":'en',
    "locale": "US",
    "deviceName":"127.0.0.1:62001"

}
url = "http://127.0.0.1:4723"
driver = webdriver.Remote(command_executor=url,desired_capabilities=cap)
time.sleep(2)
driver.find_element(AppiumBy.XPATH,'//android.widget.FrameLayout[@content-desc="Folder: Tools"]').click()
time.sleep(2)
driver.find_element(AppiumBy.XPATH,"/hierarchy/android.widget.FrameLayout/android.widget.LinearLayout/android.widget.FrameLayout/android.widget.FrameLayout/android.widget.FrameLayout/android.widget.LinearLayout/com.android.launcher3.launcher3.folder.FolderPagedView/android.view.ViewGroup/android.view.View").click()
time.sleep(2)
driver.find_element(AppiumBy.XPATH,"/hierarchy/android.widget.FrameLayout/android.widget.LinearLayout/android.widget.FrameLayout/android.support.v4.widget.DrawerLayout/android.widget.LinearLayout/android.widget.FrameLayout[2]/android.widget.FrameLayout/android.support.v4.view.ViewPager/android.support.v7.widget.RecyclerView/android.widget.LinearLayout[5]/android.widget.LinearLayout").click()
el1 = driver.find_element(AppiumBy.XPATH,"//*[@text='Wi‑Fi']")
el2 = driver.find_element(AppiumBy.XPATH,"//*[@text='Apps']")
driver.scroll(el2,el1)
driver.swipe(start_x=200,start_y=1870,end_x=200,end_y=200)
driver.quit()
`
and the log

[debug] [HTTP] Request idempotency key: 66afaf10-5b13-4a30-a54a-9c4718c83f9f
[HTTP] --> POST /session
[HTTP] {"capabilities":{"firstMatch":[{}],"alwaysMatch":{"platformName":"Android","appium:automationName":"uiautomator2","appium:appPackage":"com.android.launcher3","appium:appActivaty":".launcher3.Launcher","appium:language":"en","appium:locale":"US","appium:deviceName":"127.0.0.1:62001"}}}
[debug] [AppiumDriver@43cb] Calling AppiumDriver.createSession() with args: [null,null,{"firstMatch":[{}],"alwaysMatch":{"platformName":"Android","appium:automationName":"uiautomator2","appium:appPackage":"com.android.launcher3","appium:appActivaty":".launcher3.Launcher","appium:language":"en","appium:locale":"US","appium:deviceName":"127.0.0.1:62001"}}]
[debug] [AppiumDriver@43cb] Event 'newSessionRequested' logged at 1689750570786 (15:09:30 GMT+0800 (中国标准时间))
[Appium] Attempting to find matching driver for automationName 'uiautomator2' and platformName 'Android'
[Appium] The 'uiautomator2' driver was installed and matched caps.
[Appium] Will require it at C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver
[debug] [Appium] Requiring driver at C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver
[AppiumDriver@43cb] Appium v2.0.0-rc.5 creating new AndroidUiautomator2Driver (v2.29.2) session
[AppiumDriver@43cb] Checking BaseDriver versions for Appium and AndroidUiautomator2Driver
[AppiumDriver@43cb] Appium's BaseDriver version is 9.3.15
[AppiumDriver@43cb] AndroidUiautomator2Driver's BaseDriver version is 9.3.15
[debug] [AndroidUiautomator2Driver@087c] Creating session with W3C capabilities: {
[debug] [AndroidUiautomator2Driver@087c]   "alwaysMatch": {
[debug] [AndroidUiautomator2Driver@087c]     "platformName": "Android",
[debug] [AndroidUiautomator2Driver@087c]     "appium:automationName": "uiautomator2",
[debug] [AndroidUiautomator2Driver@087c]     "appium:appPackage": "com.android.launcher3",
[debug] [AndroidUiautomator2Driver@087c]     "appium:appActivaty": ".launcher3.Launcher",
[debug] [AndroidUiautomator2Driver@087c]     "appium:language": "en",
[debug] [AndroidUiautomator2Driver@087c]     "appium:locale": "US",
[debug] [AndroidUiautomator2Driver@087c]     "appium:deviceName": "127.0.0.1:62001"
[debug] [AndroidUiautomator2Driver@087c]   },
[debug] [AndroidUiautomator2Driver@087c]   "firstMatch": [
[debug] [AndroidUiautomator2Driver@087c]     {}
[debug] [AndroidUiautomator2Driver@087c]   ]
[debug] [AndroidUiautomator2Driver@087c] }
[AndroidUiautomator2Driver@087c] The following capabilities were provided, but are not recognized by Appium:
[AndroidUiautomator2Driver@087c]   appActivaty
[AndroidUiautomator2Driver@087c]   deviceName
[AndroidUiautomator2Driver@087c (2967b70c)] Session created with session id: 2967b70c-f0e3-40ae-8f49-bfc40828a1ae
[AndroidUiautomator2Driver@087c (2967b70c)] Starting 'com.android.launcher3' directly on the device
[ADB] Found 1 'build-tools' folders under 'C:\Users\BBG\AppData\Local\Android\Sdk' (newest first):
[ADB]     C:/Users/BBG/AppData/Local/Android/Sdk/build-tools/34.0.0
[ADB] Using 'adb.exe' from 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 start-server'
[AndroidDriver] Retrieving device list
[debug] [ADB] Trying to find a connected android device
[debug] [ADB] Getting connected devices
[debug] [ADB] Connected devices: [{"udid":"127.0.0.1:62001","state":"device"},{"udid":"127.0.0.1:62025","state":"device"}]
[AndroidDriver] Using device: 127.0.0.1:62001
[ADB] Using 'adb.exe' from 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 start-server'
[debug] [ADB] Setting device id to 127.0.0.1:62001
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell getprop ro.build.version.sdk'
[debug] [ADB] Current device property 'ro.build.version.sdk': 25
[debug] [ADB] Device API level: 25
[AndroidDriver] No app sent in, not parsing package/activity
[debug] [AndroidDriver] Pushing settings apk to device...
[debug] [ADB] Getting package info for 'io.appium.settings'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell dumpsys package io.appium.settings'
[debug] [ADB] The version name of the installed 'io.appium.settings' is greater or equal to the application version name ('5.0.5' >= '5.0.5')
[debug] [ADB] There is no need to install/upgrade 'C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\io.appium.settings\apks\settings_apk-debug.apk'
[debug] [ADB] Getting IDs of all 'io.appium.settings' processes
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1\:62001 shell 'pgrep --help; echo $?''
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1\:62001 shell 'pgrep ^appium\\.settings$ || pgrep ^io\\.appium\\.setti$''
[debug] [AndroidDriver] io.appium.settings is already running. There is no need to reset its permissions.
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell appops set io.appium.settings android:mock_location allow'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1\:62001 shell "[ -e '/data/local/tmp/mock_apps.json' ] && echo __PASS__"'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell cat /data/local/tmp/mock_apps.json'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell getprop persist.sys.locale'
[debug] [ADB] Current device property 'persist.sys.locale': en-US
[debug] [ADB] Current locale: 'en-US'; requested locale: 'en-US'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell getprop persist.sys.locale'
[debug] [ADB] Current device property 'persist.sys.locale': en-US
[debug] [ADB] Requested locale is equal to current locale: 'en-us'
[debug] [Logcat] Starting logs capture with command: C:\\Users\\BBG\\AppData\\Local\\Android\\Sdk\\platform-tools\\adb.exe -P 5037 -s 127.0.0.1\:62001 logcat -v threadtime
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Forwarding UiAutomator2 Server port 6790 to local port 8202
[debug] [ADB] Forwarding system: 8202 to device: 6790
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 forward tcp:8202 tcp:6790'
[debug] [ADB] Checking app cert for C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\appium-uiautomator2-server\apks\appium-uiautomator2-server-v5.12.2.apk
[debug] [ADB] Checking app cert for C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\appium-uiautomator2-server\apks\appium-uiautomator2-server-debug-androidTest.apk
[ADB] Using 'apksigner.jar' from 'C:\Users\BBG\AppData\Local\Android\Sdk\build-tools\34.0.0\lib\apksigner.jar'
[debug] [ADB] Starting apksigner: D:\\auto_app\\jdk-9+181_windows-x64_ri\\java-se-9-ri\\jdk-9\\bin\\java.exe -Xmx1024M -Xss1m -jar C:\\Users\\BBG\\AppData\\Local\\Android\\Sdk\\build-tools\\34.0.0\\lib\\apksigner.jar verify --print-certs C:\\Users\\BBG\\.appium\\node_modules\\appium-uiautomator2-driver\\node_modules\\appium-uiautomator2-server\\apks\\appium-uiautomator2-server-debug-androidTest.apk
[debug] [ADB] Starting apksigner: D:\\auto_app\\jdk-9+181_windows-x64_ri\\java-se-9-ri\\jdk-9\\bin\\java.exe -Xmx1024M -Xss1m -jar C:\\Users\\BBG\\AppData\\Local\\Android\\Sdk\\build-tools\\34.0.0\\lib\\apksigner.jar verify --print-certs C:\\Users\\BBG\\.appium\\node_modules\\appium-uiautomator2-driver\\node_modules\\appium-uiautomator2-server\\apks\\appium-uiautomator2-server-v5.12.2.apk
[debug] [ADB] apksigner stdout: Signer #1 certificate DN: EMAILADDRESS=android@android.com, CN=Android, OU=Android, O=Android, L=Mountain View, ST=California, C=US
[debug] [ADB] Signer #1 certificate SHA-256 digest: a40da80a59d170caa950cf15c18c454d47a39b26989d8b640ecd745ba71bf5dc
[debug] [ADB] Signer #1 certificate SHA-1 digest: 61ed377e85d386a8dfee6b864bd85b0bfaa5af81
[debug] [ADB] Signer #1 certificate MD5 digest: e89b158e4bcf988ebd09eb83f5378e87
[debug] [ADB]
[debug] [ADB] sha256 hash did match for 'appium-uiautomator2-server-debug-androidTest.apk'
[ADB] 'C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\appium-uiautomator2-server\apks\appium-uiautomator2-server-debug-androidTest.apk' is signed with the default certificate
[debug] [ADB] Getting install status for io.appium.uiautomator2.server.test
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell pm path io.appium.uiautomator2.server.test'
[debug] [ADB] 'io.appium.uiautomator2.server.test' is installed
[debug] [ADB] apksigner stdout: Signer #1 certificate DN: EMAILADDRESS=android@android.com, CN=Android, OU=Android, O=Android, L=Mountain View, ST=California, C=US
[debug] [ADB] Signer #1 certificate SHA-256 digest: a40da80a59d170caa950cf15c18c454d47a39b26989d8b640ecd745ba71bf5dc
[debug] [ADB] Signer #1 certificate SHA-1 digest: 61ed377e85d386a8dfee6b864bd85b0bfaa5af81
[debug] [ADB] Signer #1 certificate MD5 digest: e89b158e4bcf988ebd09eb83f5378e87
[debug] [ADB]
[debug] [ADB] sha256 hash did match for 'appium-uiautomator2-server-v5.12.2.apk'
[ADB] 'C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\appium-uiautomator2-server\apks\appium-uiautomator2-server-v5.12.2.apk' is signed with the default certificate
[debug] [ADB] Getting package info for 'io.appium.uiautomator2.server'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell dumpsys package io.appium.uiautomator2.server'
[debug] [ADB] The version name of the installed 'io.appium.uiautomator2.server' is greater or equal to the application version name ('5.12.2' >= '5.12.2')
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Server packages status: [{"wasSigned":true,"installState":"sameVersionInstalled","appPath":"C:\\Users\\BBG\\.appium\\node_modules\\appium-uiautomator2-driver\\node_modules\\appium-uiautomator2-server\\apks\\appium-uiautomator2-server-v5.12.2.apk","appId":"io.appium.uiautomator2.server"},{"wasSigned":true,"installState":"sameVersionInstalled","appPath":"C:\\Users\\BBG\\.appium\\node_modules\\appium-uiautomator2-driver\\node_modules\\appium-uiautomator2-server\\apks\\appium-uiautomator2-server-debug-androidTest.apk","appId":"io.appium.uiautomator2.server.test"}]
[AndroidUiautomator2Driver@087c (2967b70c)] Server packages are not going to be (re)installed
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Waiting up to 30000ms for services to be available
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell pm list instrumentation'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Instrumentation target 'io.appium.uiautomator2.server.test/androidx.test.runner.AndroidJUnitRunner' is available
[ADB] Adding packages ["io.appium.settings","io.appium.uiautomator2.server","io.appium.uiautomator2.server.test"] to Doze whitelist
[debug] [ADB] Got the following command chunks to execute: [["dumpsys","deviceidle","whitelist","+io.appium.settings",";","dumpsys","deviceidle","whitelist","+io.appium.uiautomator2.server",";","dumpsys","deviceidle","whitelist","+io.appium.uiautomator2.server.test",";"]]
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell dumpsys deviceidle whitelist +io.appium.settings ; dumpsys deviceidle whitelist +io.appium.uiautomator2.server ; dumpsys deviceidle whitelist +io.appium.uiautomator2.server.test ;'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] No app capability. Assuming it is already on the device
[debug] [ADB] Getting install status for com.android.launcher3
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell pm path com.android.launcher3'
[debug] [ADB] 'com.android.launcher3' is installed
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell am force-stop com.android.launcher3'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell pm clear com.android.launcher3'
[debug] [AndroidDriver] Performed fast reset on the installed 'com.android.launcher3' application (stop and clear)
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Performing shallow cleanup of automation leftovers
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] No obsolete sessions have been detected
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell am force-stop io.appium.uiautomator2.server.test'
[AndroidUiautomator2Driver@087c (2967b70c)] Starting UIAutomator2 server 5.12.2
[AndroidUiautomator2Driver@087c (2967b70c)] Using UIAutomator2 server from 'C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\appium-uiautomator2-server\apks\appium-uiautomator2-server-v5.12.2.apk' and test from 'C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\node_modules\appium-uiautomator2-server\apks\appium-uiautomator2-server-debug-androidTest.apk'
[AndroidUiautomator2Driver@087c (2967b70c)] Waiting up to 30000ms for UiAutomator2 to be online...
[debug] [ADB] Creating ADB subprocess with args: ["-P","5037","-s","127.0.0.1:62001","shell","am","instrument","-w","-e","disableAnalytics",true,"io.appium.uiautomator2.server.test/androidx.test.runner.AndroidJUnitRunner"]
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Matched '/status' to command name 'getStatus'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Proxying [GET /status] to [GET http://127.0.0.1:8202/status] with no body
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Got response with status 200: {"sessionId":"None","value":{"message":"UiAutomator2 Server is ready to accept commands","ready":true}}
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] The initialization of the instrumentation process took 52ms
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Matched '/session' to command name 'createSession'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Proxying [POST /session] to [POST http://127.0.0.1:8202/session] with body: {"capabilities":{"firstMatch":[{"platform":"LINUX","webStorageEnabled":false,"takesScreenshot":true,"javascriptEnabled":true,"databaseEnabled":false,"networkConnectionEnabled":true,"locationContextEnabled":false,"warnings":{},"desired":{"platformName":"Android","automationName":"uiautomator2","appPackage":"com.android.launcher3","appActivaty":".launcher3.Launcher","language":"en","locale":"US","deviceName":"127.0.0.1:62001"},"platformName":"Android","automationName":"uiautomator2","appPackage":"com.android.launcher3","appActivaty":".launcher3.Launcher","language":"en","locale":"US","deviceName":"127.0.0.1:62001","deviceUDID":"127.0.0.1:62001"}],"alwaysMatch":{}}}
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Got response with status 200: {"sessionId":"426fba3b-a6f2-43f2-938a-e47f052b70f3","value":{"capabilities":{"firstMatch":[{"platform":"LINUX","webStorageEnabled":false,"takesScreenshot":true,"javascriptEnabled":true,"databaseEnabled":false,"networkConnectionEnabled":true,"locationContextEnabled":false,"warnings":{},"desired":{"platformName":"Android","automationName":"uiautomator2","appPackage":"com.android.launcher3","appActivaty":".launcher3.Launcher","language":"en","locale":"US","deviceName":"127.0.0.1:62001"},"platformName":"Android","automationName":"uiautomator2","appPackage":"com.android.launcher3","appActivaty":".launcher3.Launcher","language":"en","locale":"US","deviceName":"127.0.0.1:62001","deviceUDID":"127.0.0.1:62001"}],"alwaysMatch":{}},"sessionId":"426fba3b-a6f2-43f2-938a-e47f052b70f3"}}
[AndroidUiautomator2Driver@087c (2967b70c)] Determined the downstream protocol as 'W3C'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Proxying [GET /appium/device/info] to [GET http://127.0.0.1:8202/session/426fba3b-a6f2-43f2-938a-e47f052b70f3/appium/device/info] with no body
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Got response with status 200: {"sessionId":"426fba3b-a6f2-43f2-938a-e47f052b70f3","value":{"androidId":"fcd29ff9a95ba78a","apiVersion":"25","bluetooth":{"state":"OFF"},"brand":"samsung","carrierName":"CMCC","displayDensity":240,"locale":"en_US","manufacturer":"samsung","model":"SM-G955N","networks":[{"capabilities":{"SSID":null,"linkDownBandwidthKbps":1048576,"linkUpstreamBandwidthKbps":1048576,"networkCapabilities":"NET_CAPABILITY_NOT_METERED,NET_CAPABILITY_NOT_RESTRICTED,NET_CAPABILITY_NOT_VPN,NET_CAPABILITY_NOT_ROAMING,NET_CAPABILITY_INTERNET,NET_CAPABILITY_TRUSTED,NET_CAPABILITY_VALIDATED","signalStrength":-55,"transportTypes":"TRANSPORT_WIFI"},"detailedState":"CONNECTED","extraInfo":"\"WiredSSID\"","isAvailable":true,"isConnected":true,"isFailover":false,"isRoaming":false,"state":"CONNECTED","subtype":0,"subtypeName":"","type":1,"typeName":"WIFI"}],"platformVersion":"7.1.2","realDisplaySize":"720x1280","timeZone":"Asia/Shanghai"}}
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell dumpsys window'
[AndroidDriver] Screen already unlocked, doing nothing
[AndroidUiautomator2Driver@087c (2967b70c)] Starting 'com.android.launcher3/undefined and waiting for 'com.android.launcher3/undefined'
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell am start -W -S -a android.intent.action.MAIN -c android.intent.category.LAUNCHER -f 0x10200000'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Proxying [GET /appium/device/pixel_ratio] to [GET http://127.0.0.1:8202/session/426fba3b-a6f2-43f2-938a-e47f052b70f3/appium/device/pixel_ratio] with no body
[AndroidUiautomator2Driver@087c (2967b70c)] socket hang up
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Deleting UiAutomator2 session
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Deleting UiAutomator2 server session
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Matched '/' to command name 'deleteSession'
[debug] [AndroidUiautomator2Driver@087c (2967b70c)] Proxying [DELETE /] to [DELETE http://127.0.0.1:8202/session/426fba3b-a6f2-43f2-938a-e47f052b70f3] with no body
[AndroidUiautomator2Driver@087c (2967b70c)] socket hang up
[AndroidUiautomator2Driver@087c (2967b70c)] Did not get confirmation UiAutomator2 deleteSession worked; Error was: UnknownError: An unknown server-side error occurred while processing the command. Original error: Could not proxy command to the remote server. Original error: socket hang up
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 shell am force-stop com.android.launcher3'
[debug] [Instrumentation] io.appium.uiautomator2.server.test.AppiumUiAutomator2Server:
[debug] [Logcat] Stopping logcat capture
[debug] [ADB] Removing forwarded port socket connection: 8202
[debug] [ADB] Running 'C:\Users\BBG\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s 127.0.0.1:62001 forward --remove tcp:8202'
[debug] [AppiumDriver@43cb] Event 'newSessionStarted' logged at 1689750590110 (15:09:50 GMT+0800 (中国标准时间))
[debug] [AppiumDriver@43cb] Encountered internal error running command: UnknownError: An unknown server-side error occurred while processing the command. Original error: Could not proxy command to the remote server. Original error: socket hang up
[debug] [AppiumDriver@43cb]     at UIA2Proxy.command (C:\Users\BBG\AppData\Roaming\npm\node_modules\appium\node_modules\@appium\base-driver\lib\jsonwp-proxy\proxy.js:355:13)
[debug] [AppiumDriver@43cb]     at processTicksAndRejections (node:internal/process/task_queues:95:5)
[debug] [AppiumDriver@43cb]     at AndroidUiautomator2Driver.commands.getDevicePixelRatio (C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\lib\commands\viewport.js:14:10)
[debug] [AppiumDriver@43cb]     at AndroidUiautomator2Driver.fillDeviceDetails (C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\lib\driver.js:256:28)
[debug] [AppiumDriver@43cb]     at AndroidUiautomator2Driver.createSession (C:\Users\BBG\.appium\node_modules\appium-uiautomator2-driver\lib\driver.js:242:7)
[debug] [AppiumDriver@43cb]     at AppiumDriver.createSession (C:\Users\BBG\AppData\Roaming\npm\node_modules\appium\lib\appium.js:346:35)
[HTTP] <-- POST /session 500 19856 ms - 1199

**Comment by jlipps** [comment]
You'll want to use the [--default-capabilities](http://appium.io/docs/en/2.0/cli/args/) server arg. It has replaced options like `-bp` in Appium 2. You can get some inspiration for the capabilities needed to link one Appium server to one emulator and another Appium server to another emulator in our [Grid guide](http://appium.io/docs/en/2.0/guides/grid/) (you can ignore the bits about Selenium grid specifically if you're not using that).

**Reply by 9iuzhongxian** [reply]
You have done me a great service, and many thanks for your kind and warm help.(●ˇ∀ˇ●)
